File: sorting and searching/quicksort.py
#Time complexity of Quick sort is O(n^2)

def partition(arr,si,ei):
    pivot=arr[ei]
    pindex=si
    for i in range(si,ei):
        if(arr[i]<=pivot):
            arr[i],arr[pindex]=arr[pindex],arr[i]
            pindex+=1
    arr[pindex],arr[ei]=arr[ei],arr[pindex]
    return pindex
def quicksort(arr,si,ei):
    if(si>=ei):
        return 
    pi=partition(arr,si,ei)
    quicksort(arr,si,pi-1)
    quicksort(arr,pi+1,ei)
n=int(input("Enter the size of the array:"))
print("Enter the elements of the array: ")
arr=[]
for i in range(n):
    ele=int(input())
    arr.append(ele)
quicksort(arr,0,n-1)
for i in range(n):
    print(arr[i],end=" ")

File: sorting and searching/quicksort.cpp
#include<iostream>
using namespace std;
int partition(int arr[],int si,int ei){
    int pivot=arr[ei];
    int pindex=si;//let set pindex i.e partition index to start index
    for(int i=si;i<ei;i++){
        if(arr[i]<=pivot){
            swap(arr[i],arr[pindex]);
            pindex+=1;
        }
    }
    swap(arr[pindex],arr[ei]);
    return pindex;
}
void quicksort(int arr[],int si,int ei){
    //si is start index and ei is end index
    if(si>=ei)
        return ;
    int pi=partition(arr,si,ei); //pi is partition index
    quicksort(arr,si,pi-1);
    quicksort(arr,pi+1,ei);
} 
int main(){
    int i,n;
    cout<<"Enter the size of the array :";
    cin>>n;
    cout<<"Enter the elements of the array :";
    int arr[n];
    for(i=0;i<n;i++)
        cin>>arr[i];
    quicksort(arr,0,n-1);
    cout<<"Sorted array:";
    for(i=0;i<n;i++)
        cout<<arr[i]<<" ";
}
